### ISSUE TITLE
[checkpoint] Restoring training state doesn't check whether we're fitting

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/pytorch-lightning#9404]
### Motivation

Optimizer & LR scheduler states only need to be restored during fitting. However, these are currently called unconditionally here in the checkpoint connector: 

https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L160-L161

https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L165

https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L1010-L1011

Which means we attempt to restore these states during any of `validate`/`test`/`predict`

### Pitch

Check if the trainer is fitting, and only if so restore the precision, optimizer, and LR scheduler states

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by carmocca** [comment]
This is rare but what if you have an existing checkpoint

```python
trainer.validate(ckpt_path)  # check the validation score because why not
trainer.fit()
```

In this case, you'd want to still load the optimizer state for later.

**Comment by ananthsub** [comment]
@carmocca loading the optimizer state in validate doesn't take effect today because the optimizers/LR schedulers are never set up outside of fitting: https://github.com/PyTorchLightning/pytorch-lightning/blob/089ae9b3e82ddc31942e315294e31e48c0a899db/pytorch_lightning/accelerators/accelerator.py#L285-L298

So there's no objects for the state to be restored into. This issue would make it more explicit in the codebase that this shouldn't happen by design
